File: auntieos-admin/src/api/emergencyContacts.test.ts
import { describe, it, expect, vi, beforeEach } from 'vitest';

const { call } = vi.hoisted(() => ({ call: vi.fn() }));
vi.mock('../lib/fns', () => ({ call }));

import {
  EMERGENCY_CONTACT_OUTSIDE,
  EMERGENCY_CONTACT_REQUIRED,
  draftsEqual,
  emergencyContactsOf,
  hasEmergencyContact,
  isBlankDrafts,
  listEmergencyContacts,
  saveEmergencyContacts,
  toDrafts,
  validateEmergencyContactDrafts,
} from './emergencyContacts';

beforeEach(() => call.mockReset());

const HOUSEHOLD = { names: ['Dana Mercer'], phones: ['(805) 555-0100'] };

describe('emergencyContactsOf', () => {
  it('reads the array in call order, with a Timestamp recordedAt as ISO', () => {
    const ts = { toDate: () => new Date('2026-03-02T10:00:00.000Z') };
    const out = emergencyContactsOf({
      emergencyContacts: [
        { name: 'Rae Mercer', phone: '+18055550199', relationship: 'Sister', recordedAt: ts, updatedAt: ts },
        { name: 'Lee Park', phone: '+18055550177', relationship: null },
      ],
    });
    expect(out.map((c) => c.name)).toEqual(['Rae Mercer', 'Lee Park']);
    expect(out[0]?.recordedAt).toBe('2026-03-02T10:00:00.000Z');
    expect(out[1]?.relationship).toBeNull();
  });

  it('falls back to the legacy flat triple until the migration is verified', () => {
    expect(emergencyContactsOf({ emergencyContactName: 'Rae', emergencyContactPhone: '805', emergencyContactRelation: '' })).toEqual([
      { name: 'Rae', phone: '805', relationship: null, recordedAt: null, updatedAt: null },
    ]);
    expect(hasEmergencyContact({ firstName: 'Dana' })).toBe(false);
    expect(hasEmergencyContact({ emergencyContacts: [], emergencyContactPhone: '805' })).toBe(true);
  });
});

describe('drafts', () => {
  it('toDrafts always yields at least one slot, and blank detection ignores whitespace', () => {
    expect(toDrafts([])).toEqual([{ name: '', phone: '', relationship: '' }]);
    expect(isBlankDrafts([{ name: ' ', phone: '', relationship: '' }])).toBe(true);
    expect(draftsEqual([{ name: 'A', phone: '1', relationship: '' }], [{ name: 'A', phone: '1', relationship: '' }])).toBe(true);
  });
});

describe('validateEmergencyContactDrafts', () => {
  it('requires one', () => {
    expect(validateEmergencyContactDrafts([], HOUSEHOLD)).toBe(EMERGENCY_CONTACT_REQUIRED);
    expect(validateEmergencyContactDrafts([{ name: '', phone: '', relationship: '' }], HOUSEHOLD)).toBe(EMERGENCY_CONTACT_REQUIRED);
  });
  it('needs a name and a phone on every slot', () => {
    expect(validateEmergencyContactDrafts([{ name: 'Rae', phone: '', relationship: '' }], HOUSEHOLD)).toBe('Each Emergency Contact needs a phone number.');
    expect(validateEmergencyContactDrafts([{ name: '', phone: '8055550199', relationship: '' }], HOUSEHOLD)).toBe('Each Emergency Contact needs a name.');
  });
  it('refuses the same phone twice and a household member', () => {
    expect(
      validateEmergencyContactDrafts([{ name: 'Rae', phone: '8055550199', relationship: '' }, { name: 'Lee', phone: '(805) 555-0199', relationship: '' }], HOUSEHOLD),
    ).toBe('The two Emergency Contacts need different phone numbers.');
    expect(validateEmergencyContactDrafts([{ name: 'Rae', phone: '+1 805 555 0100', relationship: '' }], HOUSEHOLD)).toBe(EMERGENCY_CONTACT_OUTSIDE);
    expect(validateEmergencyContactDrafts([{ name: ' dana  MERCER', phone: '8055550199', relationship: '' }], HOUSEHOLD)).toBe(EMERGENCY_CONTACT_OUTSIDE);
    expect(validateEmergencyContactDrafts([{ name: 'Rae', phone: '8055550199', relationship: '' }], HOUSEHOLD)).toBeNull();
  });
});

describe('callables', () => {
  it('list trims the id and decodes the answer', async () => {
    call.mockResolvedValue({ contacts: [{ name: 'Rae', phone: '+18055550199', relationship: null, recordedAt: null, updatedAt: null }], canEdit: true, legacy: false });
    const res = await listEmergencyContacts(' fam1 ');
    expect(call).toHaveBeenCalledWith('listEmergencyContacts', { kinfolkId: 'fam1' });
    expect(res.contacts[0]?.name).toBe('Rae');
  });

  it('save sends name, phone, relationship per slot in order, relationship cleared as null', async () => {
    call.mockResolvedValue({ contacts: [] });
    await saveEmergencyContacts('fam1', [
      { name: ' Rae ', phone: ' 8055550199 ', relationship: '' },
      { name: 'Lee', phone: '8055550177', relationship: ' Neighbour ' },
    ]);
    expect(call).toHaveBeenCalledWith('saveEmergencyContacts', {
      kinfolkId: 'fam1',
      contacts: [
        { name: 'Rae', phone: '8055550199', relationship: null },
        { name: 'Lee', phone: '8055550177', relationship: 'Neighbour' },
      ],
    });
  });

  it('an unreadable answer throws rather than reading as no contacts', async () => {
    call.mockResolvedValue({});
    await expect(listEmergencyContacts('fam1')).rejects.toThrow(/no contacts array/);
  });
});

File: auntieos-admin/android/app/src/test/java/com/tribetails/auntieos/data/model/EmergencyContactsTest.kt
package com.tribetails.auntieos.data.model

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class EmergencyContactsTest {

    private val household = listOf("Dana Mercer") to listOf("(805) 555-0100")

    @Test
    fun `reads the array in call order`() {
        val k = Kinfolk(
            emergencyContacts = listOf(
                mapOf("name" to "Rae Mercer", "phone" to "+18055550199", "relationship" to "Sister"),
                mapOf("name" to "Lee Park", "phone" to "+18055550177", "relationship" to null),
            ),
        )
        assertEquals(listOf("Rae Mercer", "Lee Park"), emergencyContactsOf(k).map { it.name })
        assertNull(emergencyContactsOf(k)[1].relationship)
    }

    @Test
    fun `falls back to the flat triple until the migration is verified`() {
        val k = Kinfolk(emergencyContactName = "Rae", emergencyContactPhone = "805", emergencyContactRelation = "")
        assertEquals(listOf(EmergencyContact("Rae", "805", null, null, null)), emergencyContactsOf(k))
        assertTrue(emergencyContactsOf(Kinfolk()).isEmpty())
    }

    @Test
    fun `a malformed stored value reads as none instead of crashing the profile`() {
        assertTrue(emergencyContactsOf(Kinfolk(emergencyContacts = "not a list")).isEmpty())
    }

    @Test
    fun `validation mirrors the server`() {
        val (names, phones) = household
        assertEquals(EMERGENCY_CONTACT_REQUIRED, validateEmergencyContactDrafts(listOf(EmergencyContactDraft()), names, phones))
        assertEquals("A household needs at least one Emergency Contact.", EMERGENCY_CONTACT_REQUIRED)
        assertEquals("An Emergency Contact needs a phone number.", validateEmergencyContactDrafts(listOf(EmergencyContactDraft(name = "Rae")), names, phones))
        assertEquals("An Emergency Contact needs a name.", validateEmergencyContactDrafts(listOf(EmergencyContactDraft(phone = "8055550199")), names, phones))
        assertEquals(
            "An Emergency Contact's name can be at most 80 characters.",
            validateEmergencyContactDrafts(listOf(EmergencyContactDraft("R".repeat(81), "8055550199")), names, phones),
        )
        assertEquals(
            "An Emergency Contact's phone number can be at most 32 characters.",
            validateEmergencyContactDrafts(listOf(EmergencyContactDraft("Rae", "8".repeat(33))), names, phones),
        )
        assertEquals(
            "A relationship can be at most 40 characters.",
            validateEmergencyContactDrafts(listOf(EmergencyContactDraft("Rae", "8055550199", "S".repeat(41))), names, phones),
        )
        assertEquals(
            "The two Emergency Contacts need different phone numbers.",
            validateEmergencyContactDrafts(listOf(EmergencyContactDraft("Rae", "8055550199"), EmergencyContactDraft("Lee", "(805) 555-0199")), names, phones),
        )
        assertEquals(EMERGENCY_CONTACT_OUTSIDE, validateEmergencyContactDrafts(listOf(EmergencyContactDraft("Rae", "+1 805 555 0100")), names, phones))
        assertEquals(EMERGENCY_CONTACT_OUTSIDE, validateEmergencyContactDrafts(listOf(EmergencyContactDraft(" dana  MERCER", "8055550199")), names, phones))
        assertNull(validateEmergencyContactDrafts(listOf(EmergencyContactDraft("Rae", "8055550199")), names, phones))
    }

    @Test
    fun `decode refuses a payload with no contacts array`() {
        val err = runCatching { decodeEmergencyContacts(emptyMap()) }.exceptionOrNull()
        assertTrue(err?.message.orEmpty().contains("no contacts array"))
    }
}
